File: src/main/java/service/Solver.java
package service;

import lombok.Builder;
import lombok.Data;
import model.Node;
import model.Puzzle;
import util.PuzzleType;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.stream.Collectors;

@Data
@Builder
public class Solver {
    private List<Node> open;
    private List<Node> closed;
    public final static Puzzle startPuzzle = Puzzle.getStartPuzzle(PuzzleType.DEFAULT);
    public final static Puzzle goalPuzzle = Puzzle.getGoalPuzzle();


    public int computeF(Node start, Puzzle goal){
        return this.computeH(start.getValue(), goal) + start.getLevel();
    }

    private int computeH(Puzzle value, Puzzle goal) {
        var temporary = 0;
        for (int i = 0; i < 4; i++) {
            for (int j = 0; j < 4; j++) {
                if(value.getRepresentation()[i][j] != goal.getRepresentation()[i][j]){
                    if(value.getRepresentation()[i][j] != -1){
                        temporary += 1;
                    }
                }
            }
        }
        return temporary;
    }

    public void resolve(Node root) {
        root.setFValue(this.computeF(root, goalPuzzle));
        this.open.add(root);
        var current = this.open.get(0);
        while (true){
            Solver.prettyPrint(current);
            if(this.computeH(current.getValue(), goalPuzzle) == 0){
                break;
            }
            var generatedChildren = current.generateChildren()
                    .stream().filter(e -> !this.isInClosed(e))
                    .collect(Collectors.toList());

            for (var child:
                    generatedChildren) {
                child.setFValue(this.computeF(child, goalPuzzle));
                this.open.add(child);
            }
            this.closed.add(current);
            Collections.sort(this.open);
            this.open.remove(0);
            if(generatedChildren.size() > 0){
                Collections.sort(generatedChildren);
                current = generatedChildren.get(0);
            }else{
                var finalCurrent = current;
                var possibilities = this.open.stream().filter(e -> e.getLevel() == finalCurrent.getLevel() - 1).sorted().toList();
                current = possibilities.get(0);
            }
        }
        System.out.println(current.getLevel());
    }

    public static void prettyPrint(Node current){
        System.out.println();
        System.out.println("  | ");
        System.out.println("  | ");
        System.out.println(" \\'/ \n");
        for (var i:
                current.getValue().getRepresentation()) {
            for (var j: i) {
                System.out.print(j + " ");
            }
            System.out.println();
        }
    }

    public boolean isInClosed(Node e){
        var result = new AtomicBoolean(false);
        this.closed.forEach(q -> {
            if(!q.isDifferent(e)){
                result.set(true);
            }
        });
        return result.get();
    }
}

File: csharp-pdp-project/FifteenPuzzle/FifteenPuzzleSolver/Services/Solver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using FifteenPuzzleSolver.Models;
using FifteenPuzzleSolver.Utils;

namespace FifteenPuzzleSolver.Services
{
    public class Solver
    {
        private Solver(List<Node> open, List<Node> closed)
        {
            Open = open ?? throw new ArgumentNullException(nameof(open));
            Closed = closed ?? throw new ArgumentNullException(nameof(closed));
        }

        internal static Solver CreateInstance()
        {
            return new Solver(new List<Node>(), new List<Node>());
        }

        private List<Node> Open { get; set; }
        private List<Node> Closed { get; }
        
        public static readonly Puzzle StartPuzzle = Puzzle.GetStartInstance();
        private static readonly Puzzle GoalPuzzle = Puzzle.GetGoalInstance();

        private static int ComputeF(Node start)
        {
            return ComputeH(start.Value) + start.Level;
        }

        private static int ComputeH(Puzzle startValue)
        {
            var temporary = 0;
            for (var i = 0; i < Constants.PuzzleSize; i++)
            {
                for (var j = 0; j < Constants.PuzzleSize; j++)
                {
                    if (startValue.Representation[i][j] == GoalPuzzle.Representation[i][j]) continue;
                    if (startValue.Representation[i][j] != -1)
                    {
                        temporary += 1;
                    }
                }
            }

            return temporary;
        }

        public  IEnumerable<Node> Resolve(Node root)
        {
            root.FValue = ComputeF(root);
            Open.Add(root);
            var current = root;
            var result = new List<Node>();
            while (true)
            {
                result.Add(current);
                if (ComputeH(current.Value) == 0)
                {
                    break;
                }

                var generatedChildren =  current.GenerateChildren()
                    .Where(e => e != null)
                    .Where(e => !IsInClosed(e))
                    .ToList();
                generatedChildren
                    .ToList()
                    .ForEach(child =>
                {
                    child.FValue = ComputeF(child);
                    Open.Add(child);
                });
                Closed.Add(current);
                Open.Sort();
                if (generatedChildren.Count > 0)
                {
                    
                    generatedChildren.Sort();
                    current = generatedChildren.First();
                    Open = Open.Where(e => e.IsDifferent(current)).ToList();
                }
                else
                {
                    Open.RemoveAt(0);
                    var possibilities = Open.Where(e => e.Level == current.Level - 1)
                        .Where(q => !IsInClosed(q))
                        .ToList();
                    possibilities.Sort();
                    current = possibilities.First();
                }
            }
            return result;
        }

        private bool IsInClosed(Node node)
        {
            var result = false;
            Closed.ForEach(e =>
            {
                if (!e.IsDifferent(node))
                {
                    result = true;
                }
            });
            return result;
        }

        public static void PrettyPrint(Node current){
            Console.WriteLine();
            Console.WriteLine("  | ");
            Console.WriteLine("  | ");
            Console.WriteLine(" \\'/ \n");
            foreach (var i in
            current.Value.Representation) {
                foreach (var j in i) {
                    Console.Write(j + " ");
                }
                Console.WriteLine();
            }
        }
    }
}